### PR TITLE
Add nullptr check for Party Leader in RemoveEntity

### LINKED ISSUE #5943: 🐛 Two separate nullptr crashes when calling RemoveEntity
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my issue will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened, and I have checked the commit log to see if the issue has been resolved since my server was last updated.

## OS / platform the server is running (if known)

Windows 11

## Branch affected by issue 

`base`

## Steps to reproduce

```
=== Call stack ===
Address           Frame             Function
00007FF7D20801DA  000000E548AFD1E0  CBattlefield::RemoveEntity+9A (C:\catseyexi_new\src\map\battlefield.cpp, line 527)
00007FF7D1D4527E  000000E548AFD210  CBaseEntity::~CBaseEntity+2E (C:\catseyexi_new\src\map\entities\baseentity.cpp, line 70)
00007FF7D1D546A4  000000E548AFD240  CCharEntity::`scalar deleting destructor'+14 ()
00007FF7D211F1C5  000000E548AFD290  CZoneEntities::~CZoneEntities+2B5 (C:\catseyexi_new\src\map\zone_entities.cpp, line 110)
00007FF7D212C745  000000E548AFD2D0  CZone::~CZone+45 (C:\catseyexi_new\src\map\zone.cpp, line 144)
00007FF7D212C6A4  000000E548AFD300  CZone::`scalar deleting destructor'+14 ()
00007FF7D2073FC7  000000E548AFD330  zoneutils::FreeZoneList+37 (C:\catseyexi_new\src\map\utils\zoneutils.cpp, line 1152)
00007FF7D209EB26  000000E548AFD360  do_final+76 (C:\catseyexi_new\src\map\map.cpp, line 421)
00007FFDDAF50EF5  000000E548AFD390  _seh_filter_exe+235 ()
00007FF7D2470771  000000E548AFD3C0  `__scrt_common_main_seh'::`1'::filt$0+16 (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 304)
00007FFDD0CEF73F  000000E548AFD430  __C_specific_handler+9F ()
00007FFDDD41292F  000000E548AFD460  __chkstk+11F ()
00007FFDDD3C2554  000000E548AFDB70  RtlRaiseException+484 ()
00007FFDDD41143E  000000E548AFE2A0  KiUserExceptionDispatcher+2E ()
00007FF7D2095E15  000000E548AFEF40  CLatentEffectContainer::ProcessLatentEffect+725 (C:\catseyexi_new\src\map\latent_effect_container.cpp, line 870)
00007FF7D20956A4  000000E548AFEF80  CLatentEffectContainer::ProcessLatentEffects+44 (C:\catseyexi_new\src\map\latent_effect_container.cpp, line 675)
00007FF7D209528A  000000E548AFF000  CLatentEffectContainer::CheckLatentsPartyJobs+4A (C:\catseyexi_new\src\map\latent_effect_container.cpp, line 493)
00007FF7D20FC12F  000000E548AFF0C0  CParty::ReloadParty+25F (C:\catseyexi_new\src\map\party.cpp, line 876)
00007FF7D20314BC  000000E548AFF1F0  charutils::ReloadParty+39C (C:\catseyexi_new\src\map\utils\charutils.cpp, line 6337)
00007FF7D1D565DE  000000E548AFF230  CCharEntity::PostTick+6E (C:\catseyexi_new\src\map\entities\charentity.cpp, line 884)
00007FF7D1D3FA90  000000E548AFF2B0  CAIContainer::Tick+370 (C:\catseyexi_new\src\map\ai\ai_container.cpp, line 484)
00007FF7D2127894  000000E548AFF930  CZoneEntities::ZoneServer+14A4 (C:\catseyexi_new\src\map\zone_entities.cpp, line 1709)
00007FF7D212F55C  000000E548AFF960  CZone::ZoneServer+1C (C:\catseyexi_new\src\map\zone.cpp, line 932)
00007FF7D212C362  000000E548AFF990  zone_server+22 (C:\catseyexi_new\src\map\zone.cpp, line 82)
00007FF7D1CECF4D  000000E548AFFA30  CTaskMgr::DoTimer+15D (C:\catseyexi_new\src\common\taskmgr.cpp, line 114)
00007FF7D1CC7506  000000E548AFFE30  main+256 (C:\catseyexi_new\src\common\kernel.cpp, line 273)
00007FF7D21952E4  000000E548AFFE70  __scrt_common_main_seh+10C (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 288)
00007FFDDBBE7344  000000E548AFFEA0  BaseThreadInitThunk+14 ()
00007FFDDD3BCC91  000000E548AFFF20  RtlUserThreadStart+21 ()
```

```
=== Call stack ===
Address           Frame             Function
00007FF63D9001DA  000000D47E4FC1E0  CBattlefield::RemoveEntity+9A (C:\catseyexi\src\map\battlefield.cpp, line 527)
00007FF63D5C527E  000000D47E4FC210  CBaseEntity::~CBaseEntity+2E (C:\catseyexi\src\map\entities\baseentity.cpp, line 70)
00007FF63D5D46A4  000000D47E4FC240  CCharEntity::`scalar deleting destructor'+14 ()
00007FF63D99F1C5  000000D47E4FC290  CZoneEntities::~CZoneEntities+2B5 (C:\catseyexi\src\map\zone_entities.cpp, line 110)
00007FF63D9AC745  000000D47E4FC2D0  CZone::~CZone+45 (C:\catseyexi\src\map\zone.cpp, line 144)
00007FF63D9AC6A4  000000D47E4FC300  CZone::`scalar deleting destructor'+14 ()
00007FF63D8F3FC7  000000D47E4FC330  zoneutils::FreeZoneList+37 (C:\catseyexi\src\map\utils\zoneutils.cpp, line 1152)
00007FF63D91EB36  000000D47E4FC360  do_final+76 (C:\catseyexi\src\map\map.cpp, line 421)
00007FFDDAF50EF5  000000D47E4FC390  _seh_filter_exe+235 ()
00007FF63DCF0771  000000D47E4FC3C0  `__scrt_common_main_seh'::`1'::filt$0+16 (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 304)
00007FFDD0CEF73F  000000D47E4FC430  __C_specific_handler+9F ()
00007FFDDD41292F  000000D47E4FC460  __chkstk+11F ()
00007FFDDD3C2554  000000D47E4FCB70  RtlRaiseException+484 ()
00007FFDDD41143E  000000D47E4FD270  KiUserExceptionDispatcher+2E ()
00007FF63D5D4E8A  000000D47E4FD2D0  CCharEntity::pushPacket+14A (C:\catseyexi\src\map\entities\charentity.cpp, line 341)
00007FF63D889A16  000000D47E4FD380  battleutils::DoRandomDealToEntity+2E6 (C:\catseyexi\src\map\utils\battleutils.cpp, line 5996)
00007FF63D689B0D  000000D47E4FDF80  CLuaBaseEntity::doRandomDeal+45D (C:\catseyexi\src\map\lua\lua_baseentity.cpp, line 13487)
00007FF63D71873A  000000D47E4FE000  sol::stack::stack_detail::eval<1,CLuaBaseEntity *,0,sol::argument_handler<sol::types<bool,CLuaBaseEntity *> > &,sol::member_function_wrapper<bool (__cdecl CLuaBaseEntity::*)(CLuaBaseEntity *),bool,CLuaBaseEntity,CLuaBaseEntity *>::caller,bool (__cdecl CLuaBaseEntity::*&)(CLuaBaseEntity *),CLuaBaseEntity &>+4A (C:\catseyexi\ext\sol\include\sol\sol.hpp, line 16524)
00007FF63D6E4591  000000D47E4FE080  sol::function_detail::upvalue_this_member_function<CLuaBaseEntity,bool (__cdecl CLuaBaseEntity::*)(CLuaBaseEntity *)>::real_call+61 (C:\catseyexi\ext\sol\include\sol\sol.hpp, line 19362)
00007FFDD3572CB0  000000D47E4FE0D0  0001:0000000000001CB0 C:\catseyexi\libluajit_64.dll ()
00007FFDD3578B6A  000000D47E4FE100  lua_pcall+AA ()
00007FF63D5A9B81  000000D47E4FE170  sol::basic_protected_function<sol::basic_reference<0>,0,sol::basic_reference<0> >::call<,CLuaBaseEntity,CLuaBaseEntity,CLuaAbility,CLuaAction>+D1 (C:\catseyexi\ext\sol\include\sol\sol.hpp, line 21002)
00007FF63D81E80A  000000D47E4FE980  luautils::OnUseAbility+8DA (C:\catseyexi\src\map\lua\luautils.cpp, line 4201)
00007FF63D5D8459  000000D47E4FEA50  CCharEntity::OnAbility+A59 (C:\catseyexi\src\map\entities\charentity.cpp, line 1735)
00007FF63D5A77F8  000000D47E4FEB30  CAbilityState::Update+E8 (C:\catseyexi\src\map\ai\states\ability_state.cpp, line 129)
00007FF63D5BF96D  000000D47E4FEBB0  CAIContainer::Tick+24D (C:\catseyexi\src\map\ai\ai_container.cpp, line 447)
00007FF63D9A7894  000000D47E4FF230  CZoneEntities::ZoneServer+14A4 (C:\catseyexi\src\map\zone_entities.cpp, line 1709)
00007FF63D9AF55C  000000D47E4FF260  CZone::ZoneServer+1C (C:\catseyexi\src\map\zone.cpp, line 932)
00007FF63D9AC362  000000D47E4FF290  zone_server+22 (C:\catseyexi\src\map\zone.cpp, line 82)
00007FF63D56CF4D  000000D47E4FF330  CTaskMgr::DoTimer+15D (C:\catseyexi\src\common\taskmgr.cpp, line 114)
00007FF63D547506  000000D47E4FF730  main+256 (C:\catseyexi\src\common\kernel.cpp, line 273)
00007FF63DA152E4  000000D47E4FF770  __scrt_common_main_seh+10C (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 288)
00007FFDDBBE7344  000000D47E4FF7A0  BaseThreadInitThunk+14 ()
00007FFDDD3BCC91  000000D47E4FF820  RtlUserThreadStart+21 ()
```

## Expected behavior

Probably missing nullptr checks. I think this PR is related to at least one of the crashes.
https://github.com/LandSandBoat/server/pull/5796

### PR DESCRIPTION
**Pull request by CatsEyeXI**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?
Safeley handles effect container logic if party leader is null. closes #5943 

## Steps to test these changes
Map server will no longer crash when hitting this condition.

### CODE REVIEW

**Review by zach2good — CHANGES_REQUESTED** [review]
While this is fixing the crash for you, it looks like it's dealing with a symptom rather than an underlying issue: Under no circumstances should a party ever exist even for a moment without a leader. If we're getting to a point where statuses/latents are being checked and a party is in a broken state, we've already gone too far.

I'd rather you went through the logic around party/entity removal and made sure parties are correctly adjusted or destroyed when a leader leaves or the whole thing is disbanded.

If it isn't done this way _I fully 100% promise you_ that this logical inconsistency will reappear somewhere else, and it might be more destructive or harder to track down when it does.

**Comment by CatsEyeXI** [comment]
> While this is fixing the crash for you, it looks like it's dealing with a symptom rather than an underlying issue: Under no circumstances should a party ever exist even for a moment without a leader. If we're getting to a point where statuses/latents are being checked and a party is in a broken state, we've already gone too far.
> 
> I'd rather you went through the logic around party/entity removal and made sure parties are correctly adjusted or destroyed when a leader leaves or the whole thing is disbanded.
> 
> If it isn't done this way _I fully 100% promise you_ that this logical inconsistency will reappear somewhere else, and it might be more destructive or harder to track down when it does.

yeah, i agree with everything you're saying... think this might be a bit over my head to tackle 'properly', but I'll see if Lox can make some time to help me get to the root cause, otherise, I'll just close this out. Thanks

File: src/map/latent_effect_container.cpp [hunk 1]
```diff
@@ -867,7 +867,15 @@ bool CLatentEffectContainer::ProcessLatentEffect(CLatentEffect& latentEffect)
                     }
                 }
 
-                for (auto* trust : static_cast<CCharEntity*>(m_POwner->PParty->GetLeader())->PTrusts)
+                auto leader = (CCharEntity*)m_POwner->PParty->GetLeader();
```
Comment (WinterSolstice8) [inline src/map/latent_effect_container.cpp:870]:
```suggestion
                auto leader = dynamic_cast<CCharEntity*>(m_POwner->PParty->GetLeader());
```

**Comment by WinterSolstice8** [comment]
This is probably an order-of-operations edge case where a party leader is null because they are not in zone or just left zone
There should have been a nullptr check here to begin with so I am ok with this

**Review by zach2good — APPROVED** [review]
We've had a talk internally, party/alliance/treasure pool is a fragile and awful nightmare and is desperately in need of nuking. This is probably the only area where we'll put best practice to the side to try and keep things running. Nobody wants to touch it 💀

File: src/map/latent_effect_container.cpp (hunk 1 above)
Comment (zach2good) [reply src/map/latent_effect_container.cpp:870]: Also, could be auto* PLeader

**Pull request merged**